### PR TITLE
Avoid transaction creation for send operations with only non-isolated operation to dispatch

### PR DESCRIPTION
**Pull request by danielmarbach**
While doing performance testing we found out that sends are still a magnitude slower compared to previous versions. The effect becomes more apparent the more latency is involved between the client and the sql server.

The main reason is that we ALWAYS wrap even single insert statements in a SQL transaction when doing non-isolated dispatches. This is not necessary. For more information see

http://stackoverflow.com/questions/1171749/what-does-a-transaction-around-a-single-statement-do

### Sql Express Local

> Send: NumberOfMessages 10000, MessageSize 1024, Concurrency 1, TimeInMs 11771
> Send: NumberOfMessages 10000, MessageSize 4096, Concurrency 1, TimeInMs 12260
> Send: NumberOfMessages 10000, MessageSize 16384, Concurrency 1, TimeInMs 12276
> Send: NumberOfMessages 10000, MessageSize 65536, Concurrency 1, TimeInMs 12108
> Send: NumberOfMessages 10000, MessageSize 262144, Concurrency 1, TimeInMs 12376
> Send: NumberOfMessages 10000, MessageSize 1048576, Concurrency 1, TimeInMs 12490
> Send: NumberOfMessages 10000, MessageSize 4194304, Concurrency 1, TimeInMs 12326
> Send: NumberOfMessages 10000, MessageSize 16777216, Concurrency 1, TimeInMs 12496

Comparison to previous version

> Send: NumberOfMessages 10000, MessageSize 1024, Concurrency 1, TimeInMs 9964
> Send: NumberOfMessages 10000, MessageSize 4096, Concurrency 1, TimeInMs 8235
> Send: NumberOfMessages 10000, MessageSize 16384, Concurrency 1, TimeInMs 7682
> Send: NumberOfMessages 10000, MessageSize 65536, Concurrency 1, TimeInMs 8247
> Send: NumberOfMessages 10000, MessageSize 262144, Concurrency 1, TimeInMs 8370
> Send: NumberOfMessages 10000, MessageSize 1048576, Concurrency 1, TimeInMs 8462
> Send: NumberOfMessages 10000, MessageSize 4194304, Concurrency 1, TimeInMs 8326
> Send: NumberOfMessages 10000, MessageSize 16777216, Concurrency 1, TimeInMs 8503

### Sql Azure S3 (100 DTUs)

image

(The above screenshot only shows 888 calls to `BeginTransaction` because I took the snapshot while sending)

> Send: NumberOfMessages 1000, MessageSize 1024, Concurrency 1, TimeInMs 140979
> Send: NumberOfMessages 1000, MessageSize 4096, Concurrency 1, TimeInMs 136097
> Send: NumberOfMessages 1000, MessageSize 16384, Concurrency 1, TimeInMs 139523
> Send: NumberOfMessages 1000, MessageSize 65536, Concurrency 1, TimeInMs 136477
> Send: NumberOfMessages 1000, MessageSize 262144, Concurrency 1, TimeInMs 135952
> Send: NumberOfMessages 1000, MessageSize 1048576, Concurrency 1, TimeInMs 139093
> Send: NumberOfMessages 1000, MessageSize 4194304, Concurrency 1, TimeInMs 136451
> Send: NumberOfMessages 1000, MessageSize 16777216, Concurrency 1, TimeInMs 138022

Comparison to previous version

> Send: NumberOfMessages 1000, MessageSize 1024, Concurrency 1, TimeInMs 33880
> Send: NumberOfMessages 1000, MessageSize 4096, Concurrency 1, TimeInMs 34149
> Send: NumberOfMessages 1000, MessageSize 16384, Concurrency 1, TimeInMs 33810
> Send: NumberOfMessages 1000, MessageSize 65536, Concurrency 1, TimeInMs 35401
> Send: NumberOfMessages 1000, MessageSize 262144, Concurrency 1, TimeInMs 33905
> Send: NumberOfMessages 1000, MessageSize 1048576, Concurrency 1, TimeInMs 34100
> Send: NumberOfMessages 1000, MessageSize 4194304, Concurrency 1, TimeInMs 34183
> Send: NumberOfMessages 1000, MessageSize 16777216, Concurrency 1, TimeInMs 33200

### Sql Express Local

> Send: NumberOfMessages 10000, MessageSize 1024, Concurrency 1, TimeInMs 7196
> Send: NumberOfMessages 10000, MessageSize 4096, Concurrency 1, TimeInMs 6411
> Send: NumberOfMessages 10000, MessageSize 16384, Concurrency 1, TimeInMs 7014
> Send: NumberOfMessages 10000, MessageSize 65536, Concurrency 1, TimeInMs 6332
> Send: NumberOfMessages 10000, MessageSize 262144, Concurrency 1, TimeInMs 7240
> Send: NumberOfMessages 10000, MessageSize 1048576, Concurrency 1, TimeInMs 7259
> Send: NumberOfMessages 10000, MessageSize 4194304, Concurrency 1, TimeInMs 7438
> Send: NumberOfMessages 10000, MessageSize 16777216, Concurrency 1, TimeInMs 7257

### Sql Azure S3 (100 DTUs)

image

The two calls are from preping for the test run

> Send: NumberOfMessages 1000, MessageSize 1024, Concurrency 1, TimeInMs 33655
> Send: NumberOfMessages 1000, MessageSize 4096, Concurrency 1, TimeInMs 32165
> Send: NumberOfMessages 1000, MessageSize 16384, Concurrency 1, TimeInMs 32484
> Send: NumberOfMessages 1000, MessageSize 65536, Concurrency 1, TimeInMs 33717
> Send: NumberOfMessages 1000, MessageSize 262144, Concurrency 1, TimeInMs 33319
> Send: NumberOfMessages 1000, MessageSize 1048576, Concurrency 1, TimeInMs 33840
> Send: NumberOfMessages 1000, MessageSize 4194304, Concurrency 1, TimeInMs 35234
> Send: NumberOfMessages 1000, MessageSize 16777216, Concurrency 1, TimeInMs 34467

That is 4 times more throughput on my tests with SQL Azure S3!

### CODE REVIEW

**Comment by danielmarbach** [comment]
@Particular/sqlserver-transport-maintainers please review

File: src/NServiceBus.SqlServer/Sending/TableBasedQueueDispatcher.cs [hunk 1]
```diff
@@ -58,14 +58,27 @@ public TableBasedQueueDispatcher(SqlConnectionFactory connectionFactory)
...
+                using (var transaction = connection.BeginTransaction())
+                {
+                    await SendInsideTransaction(defaultConsistencyOperations, connection, transaction).ConfigureAwait(false);
+                    transaction.Commit();
                 }
-                transaction.Commit();
+            }
+        }
+
+        static async Task SendInsideTransaction(List<MessageWithAddress> defaultConsistencyOperations, SqlConnection connection, SqlTransaction transaction)
```
Comment (SimonCropp) [inline src/NServiceBus.SqlServer/Sending/TableBasedQueueDispatcher.cs]: very minor but this is not really "SendInsideTransaction"

**Comment by SimonCropp** [comment]
LGTM

File: src/NServiceBus.SqlServer/Sending/TableBasedQueueDispatcher.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.SqlServer/Sending/TableBasedQueueDispatcher.cs]: I struggled to have a better name, any suggestions?

Comment (SimonCropp) [inline src/NServiceBus.SqlServer/Sending/TableBasedQueueDispatcher.cs]: perhaps just "Send"??

**Comment by danielmarbach** [comment]
@SimonCropp please have another look

**Comment by SimonCropp** [comment]
@danielmarbach looks good. @Particular/sqlserver-transport-maintainers can we get a final QA and merge please

**Pull request merged**